src/lib/operators/join_index.cpp

```
@@ -0,0 +1,346 @@
+  }
+}
+
+void JoinIndex::_perform_join() {
+  DebugAssert(_mode != JoinMode::Cross, "Cross join mode is not supported by index join");
+
+  _right_matches.resize(_right_in_table->chunk_count());
+  _left_matches.resize(_left_in_table->chunk_count());
+
+  auto left_data_type = _left_in_table->column_type(_left_column_id);
+  auto right_data_type = _right_in_table->column_type(_right_column_id);
+
+  _pos_list_left = std::make_shared<PosList>();
+  _pos_list_right = std::make_shared<PosList>();
+
+  for (ChunkID chunk_id_left = ChunkID{0}; chunk_id_left < _left_in_table->chunk_count(); ++chunk_id_left) {
+    // initialize the data structures for left matches
+    _left_matches[chunk_id_left].resize(_left_in_table->get_chunk(chunk_id_left)->size());
+  }
+
+  // Scan all chunks for right input
+  for (ChunkID chunk_id_right = ChunkID{0}; chunk_id_right < _right_in_table->chunk_count(); ++chunk_id_right) {
+    const auto chunk_right = _right_in_table->get_chunk(chunk_id_right);
+    auto column_right = chunk_right->get_column(_right_column_id);
+    const auto indices = chunk_right->get_indices(std::vector<ColumnID>{_right_column_id});
+    _right_matches[chunk_id_right].resize(chunk_right->size());
+
+    std::shared_ptr<BaseIndex> index = nullptr;
+
+    if (indices.size() > 0) {
+      // We assume the first index to be efficient for our join
+      // as we do not want to spend time on evaluating the best index inside of this join loop
+      index = indices.front();
+    }
+
+    // Scan all chunks from left input
+    for (ChunkID chunk_id_left = ChunkID{0}; chunk_id_left < _left_in_table->chunk_count(); ++chunk_id_left) {
+      auto chunk_column_left = _left_in_table->get_chunk(chunk_id_left)->get_column(_left_column_id);
+
+      // for Outer joins, remember matches on the left side
+      auto& left_matches = _left_matches[chunk_id_left];
+
+      if (_mode == JoinMode::Left || _mode == JoinMode::Outer) {
```

I think operators should not duplicate behaviour of another operator. Instead, the operator should specify which modes are supported and should fail if an unsupported mode was passed.

---

We came to the conclusion that we can support all modes without major time overhead so we did. The resizing below is erroneous and will be fixed.